larq/quantizers.py

```diff
@@ -130,6 +130,94 @@ def approx_sign(x):
     return _binarize_with_weighted_grad(x)
 
 
+@utils.register_keras_custom_object
+class SteTern:
+    r"""
+    Ternarization function.
+    \\[
+    q(x) = \begin{cases}
+    +1 & x > \Delta \\\
+    0 & |x| > \Delta \\\
+    -1 & x < \Delta
+    \end{cases}
+    \\]
+
+    where $\Delta$ is defined as the threshold and can be passed as an argument,
+    or can be calculated as per the Ternary Weight Networks original paper, such that
+
+    \\[
+    \Delta = \frac{0.7}{n} \sum_{i=1}^{n} |W_i|
+    \\]
+    where we assume that $W_i$ is generated from a normal distribution.
+
+    The gradient is estimated using the Straight-Through Estimator
+    (essentially the Ternarization is replaced by a clipped identity on the
+    backward pass).
+    \\[\frac{\partial q(x)}{\partial x} = \begin{cases}
+    1 & \left|x\right| \leq 1 \\\
+    0 & \left|x\right| > 1
+    \end{cases}\\]
+
+    ```plot-activation
+    quantizers.SteTern
+    ```
+
+    # Arguments
+    x: Input tensor.
+    threshold value: The value for the threshold, $\Delta$.
+    ternary_weight_networks: Boolean of whether to use the Ternary Weight Networks threshold calculation. 
+
+    # Returns
+    Ternarized tensor.
+
+    # References
+    - [Ternary Weight Networks](http://arxiv.org/abs/1605.04711)
+    """
+
+    def __init__(self, threshold_value=0.1, ternary_weight_networks=False):
```

Comment:
I think `ternary_weight_networks` param should be a bit more descriptive, though I can't think of a better name on top of my head.

Reply:
I'm happy with this name as it's the original paper name, and I only think they use this method in the original paper